pygradle-plugin/src/main/groovy/com/linkedin/gradle/python/tasks/ParallelWheelGenerationTask.java

```diff
@@ -0,0 +1,166 @@
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.internal.logging.progress.ProgressLogger;
+import org.gradle.internal.logging.progress.ProgressLoggerFactory;
+import org.gradle.process.ExecResult;
+
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ParallelWheelGenerationTask extends DefaultTask {
+
+    private static final Logger logger = Logging.getLogger(ParallelWheelGenerationTask.class);
+
+    private FileCollection filesToConvert;
+    private File cacheDir;
+    private PythonExtension extension;
+    private PythonDetails pythonDetails;
+    private AtomicInteger counter = new AtomicInteger();
+
+    @TaskAction
+    public void buildWheels() {
+
+        ProgressLoggerFactory progressLoggerFactory = getServices().get(ProgressLoggerFactory.class);
+        ProgressLogger progressLogger = progressLoggerFactory.newOperation(ParallelWheelGenerationTask.class);
+        progressLogger.setDescription("Building wheels");
+
+        PythonTag pythonTag = PythonTag.findTag(getProject(), getPythonDetails());
+        PlatformTag platformTag = PlatformTag.makePlatformTag(getProject(), getPythonDetails());
+
+        WheelCache wheelCache = new WheelCache(cacheDir, pythonTag, platformTag);
+
+        progressLogger.started();
+
+        TaskTimer taskTimer = new TaskTimer();
+
+        Set<File> files = getFilesToConvert().getFiles();
+        int totalSize = files.size();
+        files.parallelStream().forEach(file -> {
+            PackageInfo packageInfo = PackageInfo.fromPath(file.getPath());
+            TaskTimer.TickingClock clock = taskTimer.start(packageInfo.getName() + "-" + packageInfo.getVersion());
+            makeWheelFromSdist(progressLogger, totalSize, wheelCache, file);
+            clock.stop();
+        });
+
+        try {
+            FileUtils.write(getBuildReport(), taskTimer.buildReport());
+        } catch (IOException ignore) {
+        }
+        progressLogger.completed();
+    }
+
+    private void makeWheelFromSdist(ProgressLogger progressLogger, int totalSize, WheelCache wheelCache, File input) {
+
+        if (input.getName().endsWith(".whl")) {
+            return;
+        }
+
+        PackageInfo packageInfo = PackageInfo.fromPath(input.getPath());
+        progressLogger.progress(String.format("Building wheel %s %d of %d", packageInfo.getName(), counter.incrementAndGet(), totalSize));
+        Optional<File> cachedWheel = wheelCache.findWheel(packageInfo.getName(), packageInfo.getVersion(), pythonDetails.getPythonVersion());
+
+        if (cachedWheel.isPresent()) {
+            return;
+        }
+
+        ByteArrayOutputStream stream = new ByteArrayOutputStream();
+        ExecResult results = getProject().exec(exec -> {
+            exec.commandLine(pythonDetails.getVirtualEnvInterpreter(),
```

Comment:
Where do we set environment for this command execution?

Reply:
We don't.